src/log.c

```diff
@@ -133,6 +136,122 @@ void fi_log_init(void)
 	pid = getpid();
 }
 
+static int ofi_log_enabled(int disable, enum fi_log_level level,
+			   enum fi_log_subsys subsys)
+{
+    return ((FI_LOG_TAG(disable, level, subsys) & log_mask) ==
+	    FI_LOG_TAG(disable, level, subsys));
+}
+
+static void ofi_log(const char *buf)
+{
+    fprintf(stderr, "%s", buf);
+}
+
+static int ofi_log_ready(int disable, enum fi_log_level level,
+			 enum fi_log_subsys subsys, uint64_t *showtime);
+
+static struct fi_ops_log log_ops = {
+	.size = sizeof(struct fi_ops),
+	.enabled = ofi_log_enabled,
+	.ready = ofi_log_ready,
+	.log = ofi_log,
+};
+
+static int ofi_close_logging_fid(struct fid *fid);
+static int ofi_bind_logging_fid(struct fid *fid, struct fid *bfid,
+				uint64_t flags);
+
+static struct fi_ops ofi_logging_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = ofi_close_logging_fid,
+	.bind = ofi_bind_logging_fid,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+	.tostr = fi_no_tostr,
+	.ops_set = fi_no_ops_set,
+};
+
+static struct fid_logging log_fid = {
+	.fid = {
+		.fclass = FI_CLASS_LOG,
+		.ops = &ofi_logging_ops,
+	},
+	.ops = &log_ops,
+};
+
+static int ofi_close_logging_fid(struct fid *fid)
+{
+	/* Reset logging ops to default */
+	log_fid.ops->enabled = ofi_log_enabled;
+	log_fid.ops->ready = ofi_log_ready;
+	log_fid.ops->log = ofi_log;
+	log_fid.fid.context = NULL;
+	return 0;
+}
+
+static int ofi_logging_import(struct fid *fid)
+{
+	struct fid_logging *impfid;
+
+	if (fid->fclass != FI_CLASS_LOG)
+		return -FI_EINVAL;
+
+	impfid = container_of(fid, struct fid_logging, fid);
+	if (impfid->ops->size < sizeof(struct fi_ops_log))
+		return -FI_EINVAL;
+
+	log_fid.ops->enabled = impfid->ops->enabled;
+	log_fid.ops->ready = impfid->ops->ready;
+	log_fid.ops->log = impfid->ops->log;
+	return 0;
+}
+
+static int ofi_bind_logging_fid(struct fid *fid, struct fid *bfid,
+				uint64_t flags)
+{
+	if (flags || bfid->fclass != FI_CLASS_LOG)
+		return -FI_EINVAL;
+
+	return ofi_logging_import(bfid);
+}
+
+static int ofi_log_ready(int disable, enum fi_log_level level,
+			 enum fi_log_subsys subsys, uint64_t *showtime)
+{
+    uint64_t cur;
+
+    if (log_ops.enabled(disable, level, subsys)) {
+	    cur = ofi_gettime_ms();
+	    if (cur >= *showtime) {
+		    *showtime = cur + (uint64_t) log_interval;
+		    return true;
+	    }
+    }
+
+    return false;
+}
+
+int ofi_open_log(uint32_t version, void *attr, size_t attr_len,
+		 uint64_t flags, struct fid **fid, void *context)
+{
+	static ofi_atomic32_t opened_once;
+
+	if (FI_VERSION_LT(version, FI_VERSION(1, 13)) || attr_len)
+		return -FI_EINVAL;
+
+	if (flags)
+		return -FI_EBADFLAGS;
+
+	/* The logging subsystem can be opened only once! */
+	if (!ofi_atomic_cas_bool32(&opened_once, 0, 1))
+		return -FI_EALREADY;
+
+	log_fid.fid.context = context;
```

Comment:
Do we want to return the context back to the caller as an input argument of the logging functions?

Reply:
Usually with an imported fid, libfabric owns the context field.  The callback would return the log_fid instead.  That said, I'm okay not providing this, since we're designing the API to only allow 1 override.